File: Calculadora/ServidorCalc.java
#ClienteCalc.py

import sys, getopt
import socket
import json
import struct

DefaultHostName = "localhost"
DefaultPort = 12349
DefaultOperation = "+"
DefaultOperand1 = "5"
DefaultOperand2 = "3"
DebugMessages = False

class ComplexNumber:
    """Classe para números complexos"""
    def __init__(self, real, imag):
        self.real = real
        self.imag = imag
    
    def to_dict(self):
        return {"real": self.real, "imag": self.imag}
    
    @staticmethod
    def from_dict(data):
        return ComplexNumber(data["real"], data["imag"])
    
    @staticmethod
    def from_string(s):
        s = s.replace(' ', '').lower()
        
        if 'i' not in s:
            try:
                return ComplexNumber(float(s), 0)
            except:
                raise ValueError(f"Formato inválido: {s}")
        
        s = s.replace('i', '')
        
        if s == '':
            return ComplexNumber(0, 1)
        if s == '-':
            return ComplexNumber(0, -1)
        
        if '+' in s:
            real_str, imag_str = s.split('+')
            real = float(real_str) if real_str else 0
            imag = float(imag_str) if imag_str else 1
        elif '-' in s:
            if s[0] == '-':
                parts = s[1:].split('-')
                if len(parts) > 1:
                    real_str = '-' + parts[0]
                    imag_str = '-' + parts[1]
                else:
                    real_str = '-' + parts[0]
                    imag_str = ''
                real = float(real_str) if real_str else 0
                imag = float(imag_str) if imag_str else -1
            else:
                parts = s.split('-')
                real_str = parts[0]
                imag_str = '-' + parts[1] if len(parts) > 1 else ''
                real = float(real_str) if real_str else 0
                imag = float(imag_str) if imag_str else -1
        else:
            return ComplexNumber(0, float(s))
        
        return ComplexNumber(real, imag)
    
    def __str__(self):
        if self.imag == 0:
            return str(self.real)
        elif self.real == 0:
            return f"{self.imag}i"
        elif self.imag > 0:
            return f"{self.real} + {self.imag}i"
        else:
            return f"{self.real} - {abs(self.imag)}i"

def send_request(hostName, portNumber, request):
    print("Connecting to {} at port {}".format(hostName, portNumber))
    
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.connect((hostName, portNumber))
        
        request_text = json.dumps(request)
        request_bytes = request_text.encode('utf-8')
        
        if DebugMessages:
            print("Sending: {}".format(request_text))
        
        s.sendall(struct.pack("!I", len(request_bytes)))
        s.sendall(request_bytes)
        
        size_data = s.recv(4)
        if not size_data:
            raise Exception("No response from server")
        
        response_size = struct.unpack("!I", size_data)[0]
        
        response_data = b""
        while len(response_data) < response_size:
            chunk = s.recv(min(1024, response_size - len(response_data)))
            if not chunk:
                raise Exception("Incomplete response")
            response_data += chunk
        
        response = json.loads(response_data.decode('utf-8'))
        
        if DebugMessages:
            print("Received: {}".format(response))
        
        return response

def display_result(response):
    if response.get("status") == "success":
        result = response.get("result")
        op = response.get("operation")
        op1 = response.get("operand1")
        op2 = response.get("operand2")
        
        def format_operand(op):
            if isinstance(op, dict):
                if "real" in op and "imag" in op:
                    c = ComplexNumber.from_dict(op)
                    return str(c)
            return str(op)
        
        op1_str = format_operand(op1)
        op2_str = format_operand(op2)
        
        if result.get("type") == "complex":
            if "value" in result:
                c = ComplexNumber.from_dict(result["value"])
                result_str = str(c)
            else:
                result_str = result.get("display", str(result))
        else:
            result_str = str(result["value"])
        
        print("\n" + "=" * 50)
        print("Resultado: {} {} {} = {}".format(op1_str, op, op2_str, result_str))
        print("=" * 50 + "\n")
    
    else:
        print("\n" + "=" * 50)
        print("ERRO: {}".format(response.get('message', 'Unknown error')))
        print("=" * 50 + "\n")

def parse_operand(s):
    try:
        if '.' in s:
            return float(s)
        else:
            try:
                return int(s)
            except:
                c = ComplexNumber.from_string(s)
                return c.to_dict()
    except ValueError:
        c = ComplexNumber.from_string(s)
        return c.to_dict()

def startClient(hostName, portNumber, operand1_str, operator, operand2_str):
    print("Starting client for {} at port {}".format(hostName, portNumber))
    
    try:
        operand1 = parse_operand(operand1_str)
        operand2 = parse_operand(operand2_str)
        
        request = {
            "operation": operator,
            "operand1": operand1,
            "operand2": operand2
        }
        
        response = send_request(hostName, portNumber, request)
        display_result(response)
    
    except Exception as e:
        print("Error: {}".format(e))
        sys.exit(1)

def usage():
    print("ClienteCalc.py [--port <server port number>] [--name <server name>] [--debug] <operando1> <operador> <operando2>")
    print("\nOperações suportadas:")
    print("  Soma: +")
    print("  Subtração: -")
    print("  Multiplicação: *")
    print("\nExemplos:")
    print("  python ClienteCalc.py 5 + 3")
    print("  python ClienteCalc.py '3+2i' + '1-4i'")

def parseArguments(argv):
    print("Parsing arguments...")
    global DebugMessages
    
    try:
        opts, args = getopt.getopt(argv, "h", ["help", "port=", "name=", "debug"])
    except getopt.GetoptError as err:
        print(err)
        sys.exit(2)
    
    hostPort = DefaultPort
    hostName = DefaultHostName
    
    for opt, arg in opts:
        if opt in ("-h", "--help"):
            usage()
            sys.exit()
        elif opt in ("--port"):
            hostPort = int(arg)
        elif opt in ("--name"):
            hostName = arg
        elif opt in ("--debug"):
            DebugMessages = True
            print("Debug messages active.")
    
    if len(args) != 3:
        print("Error: Need 3 arguments: <operand1> <operator> <operand2>")
        print("Example: python ClienteCalc.py 5 + 3")
        print("Example: python ClienteCalc.py '3+2i' + '1-4i'")
        usage()
        sys.exit(1)
    
    operand1_str = args[0]
    operator = args[1]
    operand2_str = args[2]
    
    if operator not in ['+', '-', '*']:
        print("Error: Operator '{}' not supported. Use +, - or *".format(operator))
        sys.exit(1)
    
    startClient(hostName, hostPort, operand1_str, operator, operand2_str)

if __name__ == "__main__":
    parseArguments(sys.argv[1:])

File: Calculadora/ServidorCalc.py
import java.io.*;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class ServidorCalc {

    static final int DEFAULT_PORT = 12349;
    static boolean debugMessages = false;

    public static void main(String[] args) {
        int port = DEFAULT_PORT;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--port") && i + 1 < args.length) {
                port = Integer.parseInt(args[i + 1]);
                i++;
            }
            if (args[i].equals("--debug")) {
                debugMessages = true;
            }
        }

        startServer(port);
    }

    static void startServer(int port) {
        System.out.println("Starting server on port " + port);

        try (ServerSocket serverSocket = new ServerSocket(port)) {

            while (true) {
                Socket client = serverSocket.accept();
                System.out.println("New connection: " + client.getInetAddress());

                new Thread(() -> handleClient(client)).start();
            }

        } catch (IOException e) {
            System.out.println("Server error: " + e.getMessage());
        }
    }

    static void handleClient(Socket client) {
        try (DataInputStream in = new DataInputStream(client.getInputStream());
             DataOutputStream out = new DataOutputStream(client.getOutputStream())) {

            while (true) {
                // ler tamanho
                int size;
                try {
                    size = in.readInt();
                } catch (EOFException e) {
                    break;
                }

                byte[] data = new byte[size];
                in.readFully(data);

                String request = new String(data, StandardCharsets.UTF_8);

                if (debugMessages) {
                    System.out.println("Received: " + request);
                }

                String response = processRequest(request);

                byte[] responseBytes = response.getBytes(StandardCharsets.UTF_8);

                out.writeInt(responseBytes.length);
                out.write(responseBytes);
                out.flush();

                if (debugMessages) {
                    System.out.println("Sent: " + response);
                }
            }

        } catch (IOException e) {
            System.out.println("Client error: " + e.getMessage());
        }
    }

    static String processRequest(String json) {
        try {
            String op = extractString(json, "operation");
            String op1 = extractValue(json, "operand1");
            String op2 = extractValue(json, "operand2");

            if (op == null || op1 == null || op2 == null) {
                return error("Missing fields");
            }

            boolean isComplex = op1.contains("real") || op2.contains("real");

            if (isComplex) {
                Complex c1 = parseComplex(op1);
                Complex c2 = parseComplex(op2);

                Complex result;

                switch (op) {
                    case "+":
                        result = c1.add(c2);
                        break;
                    case "-":
                        result = c1.sub(c2);
                        break;
                    case "*":
                        result = c1.mul(c2);
                        break;
                    default:
                        return error("Operação inválida");
                }

                return "{"
                        + "\"status\":\"success\","
                        + "\"result\":{"
                        + "\"type\":\"complex\","
                        + "\"value\":{"
                        + "\"real\":" + result.real + ","
                        + "\"imag\":" + result.imag
                        + "},"
                        + "\"display\":\"" + result + "\""
                        + "}"
                        + "}";

            } else {
                double n1 = Double.parseDouble(op1);
                double n2 = Double.parseDouble(op2);
                double result;

                switch (op) {
                    case "+":
                        result = n1 + n2;
                        break;
                    case "-":
                        result = n1 - n2;
                        break;
                    case "*":
                        result = n1 * n2;
                        break;
                    default:
                        return error("Operação inválida");
                }

                return "{"
                        + "\"status\":\"success\","
                        + "\"result\":{"
                        + "\"type\":\"real\","
                        + "\"value\":" + result + ","
                        + "\"display\":\"" + result + "\""
                        + "}"
                        + "}";
            }

        } catch (Exception e) {
            return error("Erro: " + e.getMessage());
        }
    }

    static String error(String msg) {
        return "{ \"status\":\"error\", \"message\":\"" + msg + "\" }";
    }


    static String extractString(String json, String field) {
        String key = "\"" + field + "\"";
        int i = json.indexOf(key);
        if (i == -1) return null;

        i = json.indexOf(":", i) + 1;

        while (json.charAt(i) == ' ') i++;

        if (json.charAt(i) != '"') return null;
        i++;

        int end = json.indexOf("\"", i);
        return json.substring(i, end);
    }

    static String extractValue(String json, String field) {
        String key = "\"" + field + "\"";
        int i = json.indexOf(key);
        if (i == -1) return null;

        i = json.indexOf(":", i) + 1;

        while (json.charAt(i) == ' ') i++;

        if (json.charAt(i) == '{') {
            int braces = 1;
            int j = i + 1;
            while (braces > 0) {
                if (json.charAt(j) == '{') braces++;
                if (json.charAt(j) == '}') braces--;
                j++;
            }
            return json.substring(i, j);
        } else {
            int j = i;
            while (j < json.length() && ",}".indexOf(json.charAt(j)) == -1) j++;
            return json.substring(i, j).trim();
        }
    }

    // numeros complexos

    static class Complex {
        double real, imag;

        Complex(double r, double i) {
            real = r;
            imag = i;
        }

        Complex add(Complex o) {
            return new Complex(real + o.real, imag + o.imag);
        }

        Complex sub(Complex o) {
            return new Complex(real - o.real, imag - o.imag);
        }

        Complex mul(Complex o) {
            return new Complex(
                    real * o.real - imag * o.imag,
                    real * o.imag + imag * o.real
            );
        }

        public String toString() {
            if (imag >= 0) return real + " + " + imag + "i";
            return real + " - " + Math.abs(imag) + "i";
        }
    }

    static Complex parseComplex(String json) {
        double real = Double.parseDouble(extractValue(json, "real"));
        double imag = Double.parseDouble(extractValue(json, "imag"));
        return new Complex(real, imag);
    }
}